src/nnet3/nnet-common.cc

```diff
@@ -412,6 +412,12 @@ size_t IndexVectorHasher::operator () (
     ans += iter->n * 1619;
     ans += iter->t * 15649;
     ans += iter->x * 89809;
+	// check 'n2 > index_vector.size()' first,
+	// otherwise, 'end - n2' will be less than 'index_vector.begin()' and
+	// cause error "vector iterator + offset out of range" in STL vector.
+	if (n2 > index_vector.size() || iter >= end - n2) 
```

Comment:
make 'size' a variable (declare it next to 'ans') as it's inefficient to keep using size() in a loop.

Reply:
updated! Define 'len' as 'index_vector.size()'. Pls review again. Thanks!